File: SMStestuscomp/SMStestuscomp.m
% clear variables and close windows
clear all
close all
clc

% load data
load('uscomp.txt')

x1 = uscomp(find(uscomp(:,7)==2),1:6);
x2 = uscomp(find(uscomp(:,7)==5),1:6);

ht = hotellingstat(x1,x2);

disp('hotelling test results')
disp('test statistic')
disp(ht.t2)
disp('degrees of freedom')
disp(ht.df)
disp('p-value')
disp(ht.pval)

n1 = 15;
n2 = 10;
n  = n1+n2;
p  = 6;

x1m = mean(x1);
x2m = mean(x2);

s   = ((n1-1)*cov(x1)+(n2-1)*cov(x2))/n;
sjj = diag(s);

t   = sqrt(p*n.^2/(n1*n2*(n-p-1))*finv(0.95,p,n-p-1)*sjj);

disp('confidence interval for each variable')
disp([round(x1m-x2m-t')',round(x1m-x2m+t')'])

File: SMStestuscomp/SMStestuscomp.R
# clear variables and close windows
rm(list=ls(all=TRUE))
graphics.off()

# install and load packages
libraries = c("Hotelling")
lapply(libraries, function(x) if (!(x %in% installed.packages())) {
  install.packages(x)
})
lapply(libraries, library, quietly = TRUE, character.only = TRUE)

# setwd("C:/...") # please set your working directory
load('uscomp.rda')

#uscomp2=uscomp[is.element(uscomp$Sector,c("Energy","Manufacturing")),]
tmp          = as.character(uscomp$Sales)
tmp[65]      = "1601" # clean data error
uscomp$Sales = as.numeric(tmp)

x1           = uscomp[uscomp$Sector=="Energy",1:6]
x2           = uscomp[uscomp$Sector=="Manufacturing",1:6]

ht           = hotelling.test(x1,x2)
print(ht)

n1           = 15
n2           = 10
n            = n1+n2
p            = 6

x1m          = sapply(x1,mean)
x2m          = sapply(x2,mean)

s            = ((n1-1)*var(x1)+(n2-1)*var(x2))/n
sjj          = diag(s)

t            = sqrt(p*n^2/(n1*n2*(n-p-1))*qf(0.95,p,n-p-1)*sjj)
as.matrix(paste( "(",round(x1m-x2m-t),", ",round(x1m-x2m+t),")",sep=""),ncol=1)
